kotlinx-coroutines-core/common/src/sync/Semaphore.kt

```diff
@@ -181,11 +183,10 @@ private class SemaphoreSegment(id: Long, prev: SemaphoreSegment?): Segment<Semap
     @Suppress("NOTHING_TO_INLINE")
     inline fun cas(index: Int, expected: Any?, value: Any?): Boolean = acquirers[index].compareAndSet(expected, value)
 
-    inline fun getAndUpdate(index: Int, function: (Any?) -> Any?): Any? {
+    inline fun getAndUpdate(index: Int, value: Any?): Any? {
```

Comment:
This function need not be `inline` anymore and shall probably be renamed to `getAndSet`. Moreover, you don't need `cas` to implement it. You can directly call `acquirers[index].getAndSet(value)`, which is more efficient.